rolling-shutter/collator/batch/batch.go

```diff
@@ -0,0 +1,224 @@
+
+	"github.com/shutter-network/rolling-shutter/rolling-shutter/medley/epochid"
+)
+
+const minimumTxGas uint64 = 21000
+
+func validateGasParams(tx *txtypes.Transaction, baseFee *big.Int) error {
+	if tx.Gas() < minimumTxGas {
+		return errors.Errorf("tx gas lower than minimum (%v < %v)", tx.Gas(), minimumTxGas)
+	}
+	if tx.GasFeeCap().Cmp(tx.GasTipCap()) < 0 {
+		return errors.Errorf("gas fee cap lower than gas tip cap (%v < %v)", tx.GasFeeCap(), tx.GasTipCap())
+	}
+	if tx.GasFeeCap().Cmp(baseFee) < 0 {
+		return errors.Errorf("gas fee cap lower than header base fee (%v < %v)", tx.GasFeeCap(), baseFee)
+	}
+	return nil
+}
+
+// CalculateGasCost calculates the overall gas-cost to be deducted from the transaction
+// sender's account balance in order for the transaction to be included in the batch.
+// The deduction will be applied in the sequencers state transition function.
+// The collator has to calculate this prior to including the transaction in the batch-proposal
+// in order to check that the sender of the transaction has enough funds to pay for
+// processing of the transaction.
+func CalculateGasCost(tx *txtypes.Transaction, baseFee *big.Int) *big.Int {
+	priorityFeeGasPrice := math.BigMin(tx.GasTipCap(), new(big.Int).Sub(tx.GasFeeCap(), baseFee))
+	gasPrice := new(big.Int).Add(priorityFeeGasPrice, baseFee)
+	return new(big.Int).Mul(gasPrice, new(big.Int).SetUint64(tx.Gas()))
+}
+
+// CalculatePriorityFee calculates the part of the overall gas-cost to be deducted from the transaction
+// sender's account balance and added to the collators (coinbase) account.
+// The deduction will be applied in the sequencers state transition function.
+// The collator can calculate this prior to including the transaction in the batch-proposal
+// in order to decide which transactions should be included in the batch based on the
+// value added to the collators balance.
+func CalculatePriorityFee(tx *txtypes.Transaction, baseFee *big.Int) *big.Int {
+	priorityFeeGasPrice := math.BigMin(tx.GasTipCap(), new(big.Int).Sub(tx.GasFeeCap(), baseFee))
+	return new(big.Int).Mul(priorityFeeGasPrice, new(big.Int).SetUint64(tx.Gas()))
+}
+
+func NewPendingTx(signer txtypes.Signer, txBytes []byte) (*PendingTransaction, error) {
+	var tx txtypes.Transaction
+	err := tx.UnmarshalBinary(txBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	sender, err := signer.Sender(&tx)
+	if err != nil {
+		return nil, err
+	}
+
+	pendingTx := &PendingTransaction{
+		txBytes: txBytes,
+		tx:      &tx,
+		sender:  sender,
+	}
+	return pendingTx, nil
+}
+
+// PendingTransaction is a wrapper struct that associates additional
+// data to an incoming shutter transaction from a user.
+// It is used to keep track of sender, gas-fees and receive-time of
+// a shutter transaction.
+type PendingTransaction struct {
+	tx       *txtypes.Transaction
+	txBytes  []byte
+	sender   common.Address
+	minerFee *big.Int
+	gasCost  *big.Int
+	time     time.Time
+}
+
```

Comment:
Just get rid of the `SetReceived` method here. Instead let me pass the time with an additional `time.Now()` argument in `NewPendingTx` (i.e. not as pointer). Also the `time` field should be renamed to something more meaningful, probably `receiveTime`.

Reply:
Done